core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -817,6 +817,18 @@ class Partition(val topicPartition: TopicPartition,
       case None => localReplica.logEndOffset.messageOffset
     }
 
+    // Only actually check the HW if this is a client request (isolation level is null)
+    if (isolationLevel.isDefined && leaderEpochStartOffsetOpt.isDefined) {
+      // Wait until the HW has caught up with the start offset from this epoch
+      if (leaderEpochStartOffsetOpt.get > localReplica.highWatermark.messageOffset) {
+        throw Errors.OFFSET_NOT_AVAILABLE.exception(s"Failed to fetch offsets for " +
```

Comment:
Does this exception apply for all cases or only when the offset for latest timestamp is requested?

Reply:
That's a good question. Presumably it is possible for a timestamp query to find an offset between the log end offset and the high watermark as well. I guess the point is that we cannot trust the high watermark upper bound until we can ensure that it cannot have gone backwards. Maybe the only offset you can safely query without this validation is the earliest?

Reply:
Hmm seems like this behavior was changed then in 6a56493?

Reply:
Indeed! After talking through the various cases with @hachikuji, we decided to only enforce the check on latest timestamp offset requests.

Reply:
Thanks! Would be good to document the change in the KIP, given the following is now out-dated:
```
The KIP-207 behavior applies to all ListOffsetsRequests, whether they are for the latest offset, the earliest offset, or a time-based offset
```